coast/gridded.py

```diff
@@ -0,0 +1,865 @@
+                e3w_0_on_f = 0.25 * (e3w_0[:, :-1, :-1] + e3w_0[:, :-1, 1:] +
+                                     e3w_0[:, 1:, :-1] + e3w_0[:, 1:, 1:])
+                depth_0 = np.zeros_like(e3w_0)
+                depth_0[0, :-1, :-1] = 0.5 * e3w_0_on_f[0, :, :]
+                depth_0[1:, :-1, :-1] = depth_0[0, :-1, :-1] + np.cumsum(e3w_0_on_f[1:, :, :], axis=0)
+                bathymetry[:-1, :-1] = 0.25 * (bathymetry[:-1, :-1] + bathymetry[:-1, 1:]
+                                               + bathymetry[1:, :-1] + bathymetry[1:, 1:])
+            else:
+                raise ValueError(str(self) + ": " + self.grid_ref + " depth calculation not implemented")
+            # Write the depth_0 variable to the domain_dataset DataSet, with grid type
+            dataset_domain[f"depth{self.grid_ref.replace('-grid', '')}_0"] = xr.DataArray(
+                depth_0,
+                dims=['z_dim', 'y_dim', 'x_dim'],
+                attrs={'units': 'm', 'standard_name': 'Depth at time zero on the {}'.format(self.grid_ref)})
+
+            self.dataset['bathymetry'] = bathymetry
+            self.dataset['bathymetry'].attrs = {'units': 'm', 'standard_name': 'bathymetry',
+                                                'description': 'depth of last wet w-level on the horizontal {}'.format(
+                                                    self.grid_ref)}
+        except ValueError as err:
+            error(err)
+
+    # Add subset method to NEMO class
+    def subset_indices(self, start: tuple, end: tuple) -> tuple:
+        """
+        based on transect_indices, this method looks to return all indices between the given points.
+        This results in a 'box' (Quadrilateral) of indices.
+        consequently the returned lists may have different lengths.
+        :param start: A lat/lon pair
+        :param end: A lat/lon pair
+        :return: list of y indices, list of x indices,
+        """
+        debug(f"Subsetting {get_slug(self)} indices from {start} to {end}")
+        [j1, i1] = self.find_j_i(start[0], start[1])  # lat , lon
+        [j2, i2] = self.find_j_i(end[0], end[1])  # lat , lon
+
+        return list(np.arange(j1, j2 + 1)), list(np.arange(i1, i2 + 1))
+
+    def find_j_i(self, lat: float, lon: float):
+        """
+        A routine to find the nearest y x coordinates for a given latitude and longitude
+        Usage: [y,x] = find_j_i(49, -12)
+
+        :param lat: latitude
+        :param lon: longitude
+        :return: the y and x coordinates for the NEMO object's grid_ref, i.e. t,u,v,f,w.
+        """
+        debug(f"Finding j,i for {lat},{lon} from {get_slug(self)}")
+        dist2 = xr.ufuncs.square(self.dataset.latitude - lat) + xr.ufuncs.square(self.dataset.longitude - lon)
+        [y, x] = np.unravel_index(dist2.argmin(), dist2.shape)
+        return [y, x]
+
+    def find_j_i_domain(self, lat: float, lon: float, dataset_domain: xr.DataArray):
```

Comment:
This seems to be be a significant amount of duplication from NEMO. How much could just be moved to a common base class?

Reply:
NEMO will be remove when Gridded passes all the checks